src/vg-tooltip.js

```diff
@@ -0,0 +1,570 @@
+"use strict";
+
+(function() {
+  /**
+  * Export Vega Tooltip API: vg.tooltip(vgView, options)
+  * options can specify to show all fields or to show custom fields in the tooltip
+  * and can provide titles and formats for those fields
+  */
+  window.vg.tooltip = function(vgView, options) {
+    if (!options) {
+      options = {};
+    }
+
+    // initialize tooltip with data
+    vgView.on("mouseover", function(event, item) {
+      init(event, item, options);
+    });
+
+    // update tooltip position on mouse move
+    // (important for large marks e.g. bars)
+    vgView.on("mousemove", function(event, item) {
+      update(event, item, options);
+    });
+
+    // clear tooltip
+    vgView.on("mouseout", function(event, item) {
+      clear();
+    });
+  };
+
+  /**
+  * Export Vega-Lite Tooltip API: vl.tooltip(vgView, vlSpec, options)
+  * options can specify to show all fields or to show custom fields in the tooltip
+  * we use vlSpec to supplement options
+  */
+  window.vl.tooltip = function(vgView, vlSpec, options) {
+    if (!options) {
+      options = {};
+    }
+
+    options = supplementOptions(options, vlSpec);
+
+    // initialize tooltip with data
+    vgView.on("mouseover", function(event, item) {
+      init(event, item, options);
+    });
+
+    // update tooltip position on mouse move
+    // (important for large marks e.g. bars)
+    vgView.on("mousemove", function(event, item) {
+      update(event, item, options);
+    });
+
+    // clear tooltip
+    vgView.on("mouseout", function(event, item) {
+      clear();
+    });
+  };
+
+  /* Mapping from fieldDef.type to options.formatType */
+  var formatTypeMap = {
+    'quantitative': 'number',
+    'ordinal': undefined,
```

Comment:
`formatTypeMap` is only used to infer `formatType` from the vega-lite spec.

If a channel `'ordinal'` according can be `'number'` (e.g. 1, 2, 3...), `'string'` (e.g. A, B, C...) or `'time'` (e.g. Jan, Feb, Mar...). We cannot really infer `formatType` from the spec. So I left it as `undefined`, which means we will do auto inference later in the code.

Well I guess '`nominal`' can be `'number'`, `'string'` and `'time'` too. I will change it to `undefined`